File: src/main/kotlin/magicbook/gtlitecore/api/block/impl/TranslatableVariantItemBlock.java
package magicbook.gtlitecore.api.block.impl;

import magicbook.gtlitecore.api.block.IBlockTranslatable;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public class TranslatableVariantItemBlock<T extends Block & IBlockTranslatable> extends ItemBlock
{

    private final T block;

    public TranslatableVariantItemBlock(T block)
    {
        super(block);
        this.block = block;
        this.setHasSubtypes(true);
    }

    @Override
    public int getMetadata(int damage)
    {
        return damage;
    }

    @NotNull
    @Override
    public String getTranslationKey(@NotNull ItemStack stack)
    {
        return this.block.getTranslation(getBlockState(stack));
    }

    @SuppressWarnings("deprecation")
    public IBlockState getBlockState(ItemStack stack)
    {
        return this.block.getStateFromMeta(stack.getItemDamage());
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/block/TranslatableVariantItemBlock.kt
package gregtechlite.gtlitecore.api.block

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemStack

class TranslatableVariantItemBlock<T>(block: T) : ItemBlock(block) where T : Block, T : TranslatableBlock
{

    init
    {
        this.setHasSubtypes(true)
    }

    override fun getMetadata(damage: Int): Int = damage

    override fun getTranslationKey(stack: ItemStack): String
    {
        return (block as TranslatableBlock).getTranslation(getBlockState(stack))
    }

    @Suppress("Deprecation")
    fun getBlockState(stack: ItemStack): IBlockState
    {
        return this.block.getStateFromMeta(stack.getItemDamage())
    }

}